pera-execution/latest/pera-adapter/src/execution_engine.rs

```diff
@@ -297,6 +300,7 @@ mod checked {
 
         // We must charge object read here during transaction execution, because if this fails
         // we must still ensure an effect is committed and all objects versions incremented
```

Comment:
There is another reason you must charge: a malicious user can try to DOS the network by sending invalid transactions. Charging her regardless should resolve this.